src/tracing/metric_macros.h

```diff
@@ -0,0 +1,283 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+#include <matter/tracing/build_config.h>
+
+#define __LOG_METRIC_CONCAT_IMPL(a, b) a##b
+#define __LOG_METRIC_MACRO_CONCAT(a, b) __LOG_METRIC_CONCAT_IMPL(a, b)
+
+#if MATTER_TRACING_ENABLED
+
+/**
+ *  @def SuccessOrExitWithMetric(kMetriKey, aStatus)
+ *
+ *  @brief
+ *    This checks for the specified status, which is expected to
+ *    commonly be successful (CHIP_NO_ERROR), and branches to
+ *    the local label 'exit' if the status is unsuccessful.
+ *    If unsuccessful, a metric with key kMetriKey is emitted with
+ *    the error code as the value of the metric.
+ *
+ *  Example Usage:
+ *
+ *  @code
+ *  CHIP_ERROR TryHard()
+ *  {
+ *      CHIP_ERROR err;
+ *
+ *      err = TrySomething();
+ *      SuccessOrExitWithMetric(kMetricKey, err);
+ *
+ *      err = TrySomethingElse();
+ *      SuccessOrExitWithMetric(kMetricKey, err);
+ *
+ *  exit:
+ *      return err;
+ *  }
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted
+ *                          if the condition evaluates to false. The value
+ *                          for the metric is result of the expression aStatus.
+ *  @param[in]  aStatus     A scalar status to be evaluated against zero (0).
```

Comment:
is this true? SuccessOrExit does `::chip::ChipError::IsSuccess((aStatus))` so this is not evaluated against 0.

I find it somewhat worrysome to have different semantics between `SuccessOrExit` and `SuccessOrExitWithMetric` as they sound the same except with an extra capability.

Reply:
This is existing documentation of SuccessOrExit. I just copied it

Reply:
ah, I assumed it was the error case because the signature is `LogMetricIfError(MetricKey metricKey, const ::chip::ChipError & err)`.

The issue here is there is  a conflict:
  - doc and macro says `aStatus` is a scalar value evaluated against 0 (or generally a bool in our usage)
  - LogMetricIfError takes a ChipError argument
  
  I believe this macro implementation does not work because of that.

Reply:
The comment in successOrExit seems wrong. We may want to correct that.

here aStatus **must** be a chiperror.